File: packages/contracts/src/reliability.ts
/**
 * Reliability & Calibration Contracts
 * Reference: project.md §7.1
 *
 * Every prediction must ship with reliability metrics.
 */

import { ConfidenceLevel, Timestamps, TenantScoped, ArtifactRef } from './common';

// ============================================================================
// Calibration Score
// ============================================================================

export interface CalibrationScore {
  // Overall calibration (0-1, 1 is perfect)
  overall_score: number;

  // By prediction bucket
  bucket_scores: {
    predicted_probability_range: [number, number];
    actual_frequency: number;
    sample_size: number;
    calibration_error: number;
  }[];

  // Historical validation details
  validation_runs: {
    run_id: string;
    predicted_outcome: string;
    actual_outcome: string;
    prediction_date: string;
    outcome_date: string;
    was_correct: boolean;
  }[];

  // Calibration methodology
  methodology: string;
  validation_dataset_ref?: ArtifactRef;
}

// ============================================================================
// Stability Score
// ============================================================================

export interface StabilityScore {
  // Overall stability (0-1, 1 is perfectly stable)
  overall_score: number;

  // Variance across seeds
  seed_variance: number;

  // Per-metric stability
  metric_stability: {
    metric_name: string;
    mean_value: number;
    std_deviation: number;
    coefficient_of_variation: number;
    is_stable: boolean;
  }[];

  // Number of seeds used
  seed_count: number;

  // Outlier detection
  outlier_runs?: string[];  // run_ids that were outliers
}

// ============================================================================
// Sensitivity Summary
// ============================================================================

export interface SensitivityFactor {
  variable_path: string;
  variable_label: string;

  // Impact direction and magnitude
  impact_direction: 'positive' | 'negative' | 'nonlinear';
  impact_magnitude: number;  // Normalized 0-1

  // Elasticity (% change in outcome per % change in variable)
  elasticity?: number;

  // Confidence in this sensitivity estimate
  confidence: number;
}

export interface SensitivitySummary {
  // Top influential variables
  top_factors: SensitivityFactor[];

  // Interaction effects (if any)
  interaction_effects?: {
    variables: string[];
    interaction_type: 'amplifying' | 'dampening' | 'threshold';
    description: string;
  }[];

  // Sensitivity analysis methodology
  methodology: 'one_at_a_time' | 'global' | 'local';
  sample_size: number;
}

// ============================================================================
// Drift Status
// ============================================================================

export type DriftSeverity = 'none' | 'low' | 'medium' | 'high' | 'critical';

export interface DriftIndicator {
  indicator_name: string;

  // Current vs. baseline
  baseline_value: number;
  current_value: number;
  drift_magnitude: number;

  // Statistical test
  test_type: string;
  p_value: number;
  is_significant: boolean;

  // Impact assessment
  estimated_impact_on_predictions: 'low' | 'medium' | 'high';
}

export interface DriftStatus {
  // Overall drift assessment
  severity: DriftSeverity;
  has_drift: boolean;

  // Specific drift indicators
  indicators: DriftIndicator[];

  // When drift was detected
  detected_at?: string;

  // Recommendations
  recommendations: string[];

  // Comparison baseline
  baseline_date: string;
  baseline_dataset_ref?: ArtifactRef;
}

// ============================================================================
// Data Gaps
// ============================================================================

export interface DataGap {
  gap_id: string;

  // What's missing
  data_type: string;
  description: string;

  // Scope of the gap
  affected_regions?: string[];
  affected_segments?: string[];
  affected_time_range?: [string, string];

  // Impact assessment
  impact_severity: 'low' | 'medium' | 'high';
  impact_description: string;

  // Mitigation
  mitigation_strategy?: string;
  was_mitigated: boolean;
}

export interface DataGapsSummary {
  total_gaps: number;
  critical_gaps: number;

  gaps: DataGap[];

  // Overall data completeness
  completeness_score: number;  // 0-1

  // Recommendations
  recommendations: string[];
}

// ============================================================================
// Confidence Breakdown
// ============================================================================

export interface ConfidenceBreakdown {
  // Overall confidence
  level: ConfidenceLevel;
  score: number;  // 0-1

  // Factor contributions
  factors: {
    factor_name: string;
    contribution: number;  // -1 to 1 (negative = reduces confidence)
    weight: number;
    rationale: string;
  }[];

  // Primary reasons for confidence level
  primary_reasons: string[];

  // What would increase confidence
  improvement_suggestions: string[];
}

// ============================================================================
// Reliability Report (project.md §7.1)
// ============================================================================

export interface ReliabilityReport extends TenantScoped, Timestamps {
  // Identity
  report_id: string;

  // References (can be for a node OR a run)
  node_id?: string;
  run_id?: string;
  project_id: string;

  // Calibration
  calibration: CalibrationScore;

  // Stability
  stability: StabilityScore;

  // Sensitivity
  sensitivity: SensitivitySummary;

  // Drift
  drift: DriftStatus;

  // Data gaps
  data_gaps: DataGapsSummary;

  // Overall confidence
  confidence: ConfidenceBreakdown;

  // Versioning
  methodology_version: string;
  schema_version: string;

  // Computation metadata
  computed_at: string;
  computation_time_ms: number;
}

// ============================================================================
// Anti-Leakage Guardrails (project.md §7.2)
// ============================================================================

export interface TimeCutoffValidation {
  // The cutoff date for historical data
  cutoff_date: string;

  // Validation results
  validation_passed: boolean;

  // Any violations found
  violations: {
    data_source: string;
    data_date: string;
    violation_type: 'future_data' | 'timestamp_missing' | 'suspicious_pattern';
    description: string;
  }[];

  // Data sources checked
  sources_validated: string[];
}

export interface LeakageCheck {
  check_id: string;
  run_id: string;

  // Time cutoff validation
  time_cutoff: TimeCutoffValidation;

  // Deep research data validation
  deep_research_validated: boolean;
  untagged_sources: string[];

  // Overall result
  passed: boolean;
  warnings: string[];

  checked_at: string;
}

// ============================================================================
// Create/Update DTOs
// ============================================================================

export interface ComputeReliabilityInput {
  node_id?: string;
  run_id?: string;

  // What to compute
  include_calibration: boolean;
  include_stability: boolean;
  include_sensitivity: boolean;
  include_drift: boolean;
  include_data_gaps: boolean;

  // Computation parameters
  historical_validation_cutoff?: string;
  stability_seed_count?: number;
  sensitivity_sample_size?: number;
}

export interface UpdateCalibrationInput {
  report_id: string;
  new_validation_run: CalibrationScore['validation_runs'][0];
}

// ============================================================================
// Reliability Summary (for lists/badges)
// ============================================================================

export interface ReliabilitySummary {
  report_id: string;
  node_id?: string;
  run_id?: string;

  // Quick indicators
  confidence_level: ConfidenceLevel;
  confidence_score: number;

  calibration_score: number;
  stability_score: number;

  has_drift_warning: boolean;
  drift_severity: DriftSeverity;

  critical_data_gaps: number;

  computed_at: string;
}

// ============================================================================
// Reliability Comparison (for comparing nodes)
// ============================================================================

export interface ReliabilityComparison {
  base_node_id: string;
  compared_node_ids: string[];

  comparisons: {
    node_id: string;

    confidence_delta: number;
    calibration_delta: number;
    stability_delta: number;

    new_data_gaps: number;
    resolved_data_gaps: number;

    drift_change: 'improved' | 'worsened' | 'unchanged';
  }[];

  summary: string;
}

// ============================================================================
// PHASE 6: Reliability Integration API Types
// Reference: Phase 6 Specification
// ============================================================================

/**
 * Sensitivity result from empirical distribution analysis.
 * Phase 6: P(metric op threshold) across threshold grid.
 */
export interface Phase6SensitivityResult {
  /** Array of threshold values */
  threshold_grid: number[];
  /** P(metric op threshold) for each threshold */
  probabilities: number[];
  /** Comparison operator used */
  op: string;
  /** Metric key analyzed */
  metric_key: string;
}

/**
 * Stability result from bootstrap resampling.
 * Phase 6: Deterministic seed + 95% CI.
 */
export interface Phase6StabilityResult {
  /** Mean of bootstrap estimates */
  bootstrap_mean: number;
  /** Std of bootstrap estimates */
  bootstrap_std: number;
  /** 95% CI lower bound */
  ci_95_lower: number;
  /** 95% CI upper bound */
  ci_95_upper: number;
  /** Number of bootstrap samples */
  n_bootstrap: number;
  /** Deterministic seed hash for reproducibility */
  seed_hash: string;
}

/**
 * Drift detection result using KS statistic and PSI.
 * Phase 6: stable | warning | drifting status.
 */
export interface Phase6DriftResult {
  /** Kolmogorov-Smirnov statistic */
  ks_statistic: number;
  /** KS test p-value */
  ks_pvalue: number;
  /** Population Stability Index */
  psi: number;
  /** Drift status: stable | warning | drifting */
  drift_status: 'stable' | 'warning' | 'drifting';
  /** Number of baseline runs */
  baseline_n: number;
  /** Number of recent runs */
  recent_n: number;
  /** Baseline distribution histogram */
  baseline_histogram?: number[];
  /** Recent distribution histogram */
  recent_histogram?: number[];
  /** Histogram bin edges */
  histogram_bins?: number[];
}

/**
 * Calibration summary from latest calibration job.
 * Phase 6: Brier score and ECE.
 */
export interface Phase6CalibrationSummary {
  /** Brier score (lower is better) */
  brier_score?: number;
  /** Expected Calibration Error */
  ece?: number;
  /** Calibration method used */
  method?: string;
  /** Reference to CalibrationJob */
  calibration_job_id?: string;
}

/**
 * Audit metadata for reproducibility.
 * Phase 6: All computations are deterministic and auditable.
 */
export interface Phase6AuditMetadata {
  /** Computation timestamp */
  computed_at: string;
  /** Run IDs included in computation */
  run_ids_used: string[];
  /** Filters used */
  filters_applied: Record<string, unknown>;
  /** Seed for reproducibility */
  deterministic_seed: string;
}

/**
 * Response for GET /reliability/nodes/{node_id}/reliability/summary
 * Phase 6: Main reliability summary endpoint response.
 */
export interface Phase6ReliabilitySummaryResponse {
  /** ok | insufficient_data */
  status: 'ok' | 'insufficient_data';
  /** Total runs found */
  n_runs_total: number;
  /** Runs used after filtering */
  n_runs_used: number;

  /** Sensitivity analysis (null if insufficient_data) */
  sensitivity?: Phase6SensitivityResult;
  /** Stability analysis (null if insufficient_data) */
  stability?: Phase6StabilityResult;
  /** Drift detection (null if insufficient_data) */
  drift?: Phase6DriftResult;
  /** Calibration metrics (null if insufficient_data) */
  calibration?: Phase6CalibrationSummary;

  /** Audit trail */
  audit: Phase6AuditMetadata;
}

/**
 * Response for GET /reliability/nodes/{node_id}/reliability/detail
 * Phase 6: Detailed reliability analysis with raw data.
 */
export interface Phase6ReliabilityDetailResponse {
  /** ok | insufficient_data */
  status: 'ok' | 'insufficient_data';
  /** Total runs found */
  n_runs_total: number;
  /** Runs used after filtering */
  n_runs_used: number;

  /** Sensitivity analysis */
  sensitivity?: Phase6SensitivityResult;
  /** Stability analysis */
  stability?: Phase6StabilityResult;
  /** Drift detection */
  drift?: Phase6DriftResult;
  /** Calibration metrics */
  calibration?: Phase6CalibrationSummary;

  /** Raw metric values for custom analysis */
  raw_values?: number[];
  /** Percentile breakdown */
  percentiles?: {
    p5: number;
    p25: number;
    p50: number;
    p75: number;
    p95: number;
    mean: number;
    std: number;
    min: number;
    max: number;
  };
  /** Bootstrap sample estimates */
  bootstrap_samples?: number[];

  /** Audit trail */
  audit: Phase6AuditMetadata;
}

/**
 * Query parameters for reliability endpoints.
 * Phase 6: Shared params for summary and detail.
 */
export interface Phase6ReliabilityQueryParams {
  /** Metric key to analyze (required) */
  metric_key: string;
  /** Comparison operator: gte, lte, gt, lt, eq */
  op?: 'gte' | 'lte' | 'gt' | 'lt' | 'eq';
  /** Threshold for sensitivity at specific point */
  threshold?: number;
  /** Filter by manifest hash */
  manifest_hash?: string;
  /** Time window in days */
  window_days?: number;
  /** Minimum runs required */
  min_runs?: number;
}

File: apps/api/app/api/v1/endpoints/reliability.py
"""
PHASE 6: Reliability Integration Endpoints

Integrates Sensitivity / Stability / Drift into the Reliability module.
Computes aggregate reliability metrics across multiple runs for a node.

Reference: Phase 6 specification
"""

import hashlib
import logging
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional
from uuid import UUID

import numpy as np
from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from scipy import stats
from sqlalchemy import select, and_, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.middleware.tenant import TenantContext, get_tenant_context
from app.models.user import User
from app.models.run_outcome import RunOutcome
from app.models.calibration import CalibrationJob, CalibrationJobStatus

logger = logging.getLogger(__name__)
router = APIRouter()


# =============================================================================
# PHASE 6: Pydantic Response Schemas
# =============================================================================

class SensitivityResult(BaseModel):
    """Sensitivity analysis result - P(metric op threshold) across threshold grid."""
    threshold_grid: List[float] = Field(..., description="Array of threshold values")
    probabilities: List[float] = Field(..., description="P(metric op threshold) for each threshold")
    op: str = Field(..., description="Comparison operator used")
    metric_key: str = Field(..., description="Metric key analyzed")


class StabilityResult(BaseModel):
    """Stability analysis result from bootstrap resampling."""
    bootstrap_mean: float = Field(..., description="Mean of bootstrap estimates")
    bootstrap_std: float = Field(..., description="Std of bootstrap estimates")
    ci_95_lower: float = Field(..., description="95% CI lower bound")
    ci_95_upper: float = Field(..., description="95% CI upper bound")
    n_bootstrap: int = Field(default=200, description="Number of bootstrap samples")
    seed_hash: str = Field(..., description="Deterministic seed hash for reproducibility")


class DriftResult(BaseModel):
    """Drift detection result using KS statistic and PSI."""
    ks_statistic: float = Field(..., description="Kolmogorov-Smirnov statistic")
    ks_pvalue: float = Field(..., description="KS test p-value")
    psi: float = Field(..., description="Population Stability Index")
    drift_status: str = Field(..., description="stable | warning | drifting")
    baseline_n: int = Field(..., description="Number of baseline runs")
    recent_n: int = Field(..., description="Number of recent runs")
    baseline_histogram: Optional[List[float]] = Field(None, description="Baseline distribution histogram")
    recent_histogram: Optional[List[float]] = Field(None, description="Recent distribution histogram")
    histogram_bins: Optional[List[float]] = Field(None, description="Histogram bin edges")


class CalibrationSummary(BaseModel):
    """Calibration metrics summary from latest calibration job."""
    brier_score: Optional[float] = Field(None, description="Brier score (lower is better)")
    ece: Optional[float] = Field(None, description="Expected Calibration Error")
    method: Optional[str] = Field(None, description="Calibration method used")
    calibration_job_id: Optional[str] = Field(None, description="Reference to CalibrationJob")


class AuditMetadata(BaseModel):
    """Audit metadata for reproducibility."""
    computed_at: datetime = Field(..., description="Computation timestamp")
    run_ids_used: List[str] = Field(..., description="Run IDs included in computation")
    filters_applied: Dict[str, Any] = Field(..., description="Filters used")
    deterministic_seed: str = Field(..., description="Seed for reproducibility")


class ReliabilitySummaryResponse(BaseModel):
    """Response for GET /nodes/{node_id}/reliability/summary"""
    status: str = Field(..., description="ok | insufficient_data")
    n_runs_total: int = Field(..., description="Total runs found")
    n_runs_used: int = Field(..., description="Runs used after filtering")

    # Core metrics (null if insufficient_data)
    sensitivity: Optional[SensitivityResult] = None
    stability: Optional[StabilityResult] = None
    drift: Optional[DriftResult] = None
    calibration: Optional[CalibrationSummary] = None

    # Audit trail
    audit: AuditMetadata


class ReliabilityDetailResponse(BaseModel):
    """Response for GET /nodes/{node_id}/reliability/detail - richer curves."""
    status: str = Field(..., description="ok | insufficient_data")
    n_runs_total: int = Field(..., description="Total runs found")
    n_runs_used: int = Field(..., description="Runs used after filtering")

    # Core metrics
    sensitivity: Optional[SensitivityResult] = None
    stability: Optional[StabilityResult] = None
    drift: Optional[DriftResult] = None
    calibration: Optional[CalibrationSummary] = None

    # Extended detail
    raw_values: Optional[List[float]] = Field(None, description="Raw metric values for custom analysis")
    percentiles: Optional[Dict[str, float]] = Field(None, description="Percentile breakdown")
    bootstrap_samples: Optional[List[float]] = Field(None, description="Bootstrap sample estimates")

    # Audit trail
    audit: AuditMetadata


# =============================================================================
# PHASE 6: Computation Helpers
# =============================================================================

def compute_deterministic_seed(
    tenant_id: str,
    node_id: str,
    metric_key: str,
    threshold: Optional[float],
    manifest_hash: Optional[str],
) -> str:
    """
    Compute deterministic seed from (tenant_id, node_id, metric_key, threshold, manifest_hash).
    Returns a hash string that can be used as seed.
    """
    seed_input = f"{tenant_id}:{node_id}:{metric_key}:{threshold}:{manifest_hash}"
    return hashlib.sha256(seed_input.encode()).hexdigest()[:16]


def compute_sensitivity(
    values: List[float],
    op: str,
    threshold_grid: Optional[List[float]] = None,
    n_grid: int = 20,
) -> SensitivityResult:
    """
    Compute sensitivity: empirical P(metric op threshold) across threshold grid.

    Args:
        values: Raw metric values from runs
        op: Comparison operator (gte, lte, gt, lt, eq)
        threshold_grid: Optional explicit grid, else auto-generate
        n_grid: Number of grid points if auto-generating
    """
    if not values:
        return SensitivityResult(
            threshold_grid=[],
            probabilities=[],
            op=op,
            metric_key="",
        )

    arr = np.array(values)

    # Auto-generate threshold grid if not provided
    if threshold_grid is None:
        min_val, max_val = np.min(arr), np.max(arr)
        # Add small margin
        margin = (max_val - min_val) * 0.1 if max_val > min_val else 0.1
        threshold_grid = np.linspace(min_val - margin, max_val + margin, n_grid).tolist()

    # Compute P(metric op threshold) for each threshold
    probabilities = []
    for t in threshold_grid:
        if op == "gte":
            prob = np.mean(arr >= t)
        elif op == "gt":
            prob = np.mean(arr > t)
        elif op == "lte":
            prob = np.mean(arr <= t)
        elif op == "lt":
            prob = np.mean(arr < t)
        elif op == "eq":
            prob = np.mean(arr == t)
        else:
            prob = np.mean(arr >= t)  # Default to gte
        probabilities.append(float(prob))

    return SensitivityResult(
        threshold_grid=threshold_grid,
        probabilities=probabilities,
        op=op,
        metric_key="",  # Will be set by caller
    )


def compute_stability(
    values: List[float],
    seed_hash: str,
    n_bootstrap: int = 200,
) -> StabilityResult:
    """
    Compute stability via bootstrap resampling with deterministic seed.

    Args:
        values: Raw metric values from runs
        seed_hash: Deterministic seed hash
        n_bootstrap: Number of bootstrap samples
    """
    if not values or len(values) < 2:
        return StabilityResult(
            bootstrap_mean=0.0,
            bootstrap_std=0.0,
            ci_95_lower=0.0,
            ci_95_upper=0.0,
            n_bootstrap=n_bootstrap,
            seed_hash=seed_hash,
        )

    arr = np.array(values)
    n = len(arr)

    # Convert seed_hash to integer seed
    seed_int = int(seed_hash, 16) % (2**32)
    rng = np.random.default_rng(seed_int)

    # Bootstrap resampling
    bootstrap_estimates = []
    for _ in range(n_bootstrap):
        sample = rng.choice(arr, size=n, replace=True)
        bootstrap_estimates.append(np.mean(sample))

    bootstrap_arr = np.array(bootstrap_estimates)

    return StabilityResult(
        bootstrap_mean=float(np.mean(bootstrap_arr)),
        bootstrap_std=float(np.std(bootstrap_arr)),
        ci_95_lower=float(np.percentile(bootstrap_arr, 2.5)),
        ci_95_upper=float(np.percentile(bootstrap_arr, 97.5)),
        n_bootstrap=n_bootstrap,
        seed_hash=seed_hash,
    )


def compute_drift(
    baseline_values: List[float],
    recent_values: List[float],
    n_bins: int = 10,
) -> DriftResult:
    """
    Compute drift using KS statistic and PSI.

    Args:
        baseline_values: Values from baseline period
        recent_values: Values from recent period
        n_bins: Number of bins for PSI calculation
    """
    if not baseline_values or not recent_values:
        return DriftResult(
            ks_statistic=0.0,
            ks_pvalue=1.0,
            psi=0.0,
            drift_status="stable",
            baseline_n=len(baseline_values),
            recent_n=len(recent_values),
        )

    baseline_arr = np.array(baseline_values)
    recent_arr = np.array(recent_values)

    # KS test
    ks_stat, ks_pvalue = stats.ks_2samp(baseline_arr, recent_arr)

    # PSI calculation
    # Define bins based on combined data
    combined = np.concatenate([baseline_arr, recent_arr])
    bin_edges = np.percentile(combined, np.linspace(0, 100, n_bins + 1))
    bin_edges = np.unique(bin_edges)  # Remove duplicates

    if len(bin_edges) < 2:
        # Not enough variation for PSI
        psi = 0.0
        baseline_hist = []
        recent_hist = []
    else:
        baseline_hist, _ = np.histogram(baseline_arr, bins=bin_edges, density=True)
        recent_hist, _ = np.histogram(recent_arr, bins=bin_edges, density=True)

        # Add small epsilon to avoid division by zero
        eps = 1e-10
        baseline_hist = baseline_hist + eps
        recent_hist = recent_hist + eps

        # Normalize to proportions
        baseline_prop = baseline_hist / np.sum(baseline_hist)
        recent_prop = recent_hist / np.sum(recent_hist)

        # PSI = sum((recent - baseline) * ln(recent / baseline))
        psi = float(np.sum((recent_prop - baseline_prop) * np.log(recent_prop / baseline_prop)))

    # Determine drift status
    if psi > 0.25 or ks_pvalue < 0.01:
        drift_status = "drifting"
    elif psi > 0.10 or ks_pvalue < 0.05:
        drift_status = "warning"
    else:
        drift_status = "stable"

    return DriftResult(
        ks_statistic=float(ks_stat),
        ks_pvalue=float(ks_pvalue),
        psi=float(psi),
        drift_status=drift_status,
        baseline_n=len(baseline_values),
        recent_n=len(recent_values),
        baseline_histogram=[float(x) for x in baseline_hist] if len(baseline_hist) > 0 else None,
        recent_histogram=[float(x) for x in recent_hist] if len(recent_hist) > 0 else None,
        histogram_bins=[float(x) for x in bin_edges] if len(bin_edges) > 0 else None,
    )


async def get_calibration_summary(
    db: AsyncSession,
    tenant_id: UUID,
    node_id: UUID,
) -> Optional[CalibrationSummary]:
    """Fetch latest calibration metrics for a node."""
    query = (
        select(CalibrationJob)
        .where(
            and_(
                CalibrationJob.tenant_id == tenant_id,
                CalibrationJob.node_id == node_id,
                CalibrationJob.status == CalibrationJobStatus.SUCCEEDED.value,
            )
        )
        .order_by(CalibrationJob.finished_at.desc())
        .limit(1)
    )

    result = await db.execute(query)
    job = result.scalar_one_or_none()

    if not job:
        return None

    # Extract metrics from result_json
    metrics = job.result_json.get("metrics", {}) if job.result_json else {}

    return CalibrationSummary(
        brier_score=metrics.get("brier_score"),
        ece=metrics.get("ece"),
        method=job.config_json.get("method") if job.config_json else None,
        calibration_job_id=str(job.id),
    )


# =============================================================================
# PHASE 6: Endpoints
# =============================================================================

@router.get(
    "/nodes/{node_id}/reliability/summary",
    response_model=ReliabilitySummaryResponse,
    summary="Get Reliability Summary (PHASE 6)",
    tags=["Reliability"],
)
async def get_reliability_summary(
    node_id: UUID,
    metric_key: str = Query(..., description="Metric key to analyze (required)"),
    op: str = Query("gte", description="Comparison operator: gte, lte, gt, lt, eq"),
    threshold: Optional[float] = Query(None, description="Threshold for sensitivity at specific point"),
    manifest_hash: Optional[str] = Query(None, description="Filter by manifest hash"),
    window_days: int = Query(30, ge=1, le=365, description="Time window in days"),
    min_runs: int = Query(30, ge=1, le=1000, description="Minimum runs required"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
    tenant: TenantContext = Depends(get_tenant_context),
) -> ReliabilitySummaryResponse:
    """
    PHASE 6: Get reliability summary for a node.

    Computes sensitivity, stability, and drift metrics from run outcomes.
    Returns `status: insufficient_data` if n_runs < min_runs.

    Multi-tenant scoped. All computations are deterministic and auditable.
    """
    tenant_id = UUID(tenant.tenant_id)
    computed_at = datetime.utcnow()

    # Compute deterministic seed
    seed_hash = compute_deterministic_seed(
        tenant_id=str(tenant_id),
        node_id=str(node_id),
        metric_key=metric_key,
        threshold=threshold,
        manifest_hash=manifest_hash,
    )

    # Build query for run outcomes
    cutoff_date = datetime.utcnow() - timedelta(days=window_days)

    query = (
        select(RunOutcome)
        .where(
            and_(
                RunOutcome.tenant_id == tenant_id,
                RunOutcome.node_id == node_id,
                RunOutcome.created_at >= cutoff_date,
            )
        )
        .order_by(RunOutcome.created_at.desc())
    )

    # Add manifest_hash filter if provided
    if manifest_hash:
        query = query.where(RunOutcome.manifest_hash == manifest_hash)

    result = await db.execute(query)
    outcomes = result.scalars().all()

    n_runs_total = len(outcomes)

    # Extract metric values
    values = []
    run_ids_used = []
    for outcome in outcomes:
        metrics = outcome.metrics_json or {}
        if metric_key in metrics:
            val = metrics[metric_key]
            if isinstance(val, (int, float)):
                values.append(float(val))
                run_ids_used.append(str(outcome.run_id))

    n_runs_used = len(values)

    # Build audit metadata
    audit = AuditMetadata(
        computed_at=computed_at,
        run_ids_used=run_ids_used,
        filters_applied={
            "node_id": str(node_id),
            "metric_key": metric_key,
            "manifest_hash": manifest_hash,
            "window_days": window_days,
            "op": op,
            "threshold": threshold,
        },
        deterministic_seed=seed_hash,
    )

    # Check for insufficient data
    if n_runs_used < min_runs:
        return ReliabilitySummaryResponse(
            status="insufficient_data",
            n_runs_total=n_runs_total,
            n_runs_used=n_runs_used,
            sensitivity=None,
            stability=None,
            drift=None,
            calibration=None,
            audit=audit,
        )

    # Compute sensitivity
    sensitivity = compute_sensitivity(values, op)
    sensitivity.metric_key = metric_key

    # Compute stability
    stability = compute_stability(values, seed_hash)

    # Compute drift (split into baseline vs recent)
    # Baseline: older half, Recent: newer half
    mid_point = n_runs_used // 2
    baseline_values = values[mid_point:]  # Older runs (later in sorted list)
    recent_values = values[:mid_point]    # Newer runs
    drift = compute_drift(baseline_values, recent_values)

    # Get calibration summary
    calibration = await get_calibration_summary(db, tenant_id, node_id)

    return ReliabilitySummaryResponse(
        status="ok",
        n_runs_total=n_runs_total,
        n_runs_used=n_runs_used,
        sensitivity=sensitivity,
        stability=stability,
        drift=drift,
        calibration=calibration,
        audit=audit,
    )


@router.get(
    "/nodes/{node_id}/reliability/detail",
    response_model=ReliabilityDetailResponse,
    summary="Get Reliability Detail (PHASE 6)",
    tags=["Reliability"],
)
async def get_reliability_detail(
    node_id: UUID,
    metric_key: str = Query(..., description="Metric key to analyze (required)"),
    op: str = Query("gte", description="Comparison operator: gte, lte, gt, lt, eq"),
    threshold: Optional[float] = Query(None, description="Threshold for sensitivity at specific point"),
    manifest_hash: Optional[str] = Query(None, description="Filter by manifest hash"),
    window_days: int = Query(30, ge=1, le=365, description="Time window in days"),
    min_runs: int = Query(30, ge=1, le=1000, description="Minimum runs required"),
    n_grid: int = Query(20, ge=5, le=100, description="Number of threshold grid points"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
    tenant: TenantContext = Depends(get_tenant_context),
) -> ReliabilityDetailResponse:
    """
    PHASE 6: Get detailed reliability analysis for a node.

    Same as summary but includes:
    - raw_values: Raw metric values for custom analysis
    - percentiles: Percentile breakdown (p5, p25, p50, p75, p95)
    - bootstrap_samples: Full bootstrap sample estimates
    - Richer histogram data for drift visualization

    Multi-tenant scoped. All computations are deterministic and auditable.
    """
    tenant_id = UUID(tenant.tenant_id)
    computed_at = datetime.utcnow()

    # Compute deterministic seed
    seed_hash = compute_deterministic_seed(
        tenant_id=str(tenant_id),
        node_id=str(node_id),
        metric_key=metric_key,
        threshold=threshold,
        manifest_hash=manifest_hash,
    )

    # Build query for run outcomes
    cutoff_date = datetime.utcnow() - timedelta(days=window_days)

    query = (
        select(RunOutcome)
        .where(
            and_(
                RunOutcome.tenant_id == tenant_id,
                RunOutcome.node_id == node_id,
                RunOutcome.created_at >= cutoff_date,
            )
        )
        .order_by(RunOutcome.created_at.desc())
    )

    # Add manifest_hash filter if provided
    if manifest_hash:
        query = query.where(RunOutcome.manifest_hash == manifest_hash)

    result = await db.execute(query)
    outcomes = result.scalars().all()

    n_runs_total = len(outcomes)

    # Extract metric values
    values = []
    run_ids_used = []
    for outcome in outcomes:
        metrics = outcome.metrics_json or {}
        if metric_key in metrics:
            val = metrics[metric_key]
            if isinstance(val, (int, float)):
                values.append(float(val))
                run_ids_used.append(str(outcome.run_id))

    n_runs_used = len(values)

    # Build audit metadata
    audit = AuditMetadata(
        computed_at=computed_at,
        run_ids_used=run_ids_used,
        filters_applied={
            "node_id": str(node_id),
            "metric_key": metric_key,
            "manifest_hash": manifest_hash,
            "window_days": window_days,
            "op": op,
            "threshold": threshold,
            "n_grid": n_grid,
        },
        deterministic_seed=seed_hash,
    )

    # Check for insufficient data
    if n_runs_used < min_runs:
        return ReliabilityDetailResponse(
            status="insufficient_data",
            n_runs_total=n_runs_total,
            n_runs_used=n_runs_used,
            sensitivity=None,
            stability=None,
            drift=None,
            calibration=None,
            raw_values=None,
            percentiles=None,
            bootstrap_samples=None,
            audit=audit,
        )

    arr = np.array(values)

    # Compute sensitivity with configurable grid
    sensitivity = compute_sensitivity(values, op, n_grid=n_grid)
    sensitivity.metric_key = metric_key

    # Compute stability and capture bootstrap samples
    seed_int = int(seed_hash, 16) % (2**32)
    rng = np.random.default_rng(seed_int)
    n_bootstrap = 200
    bootstrap_samples = []
    n = len(arr)
    for _ in range(n_bootstrap):
        sample = rng.choice(arr, size=n, replace=True)
        bootstrap_samples.append(float(np.mean(sample)))

    bootstrap_arr = np.array(bootstrap_samples)
    stability = StabilityResult(
        bootstrap_mean=float(np.mean(bootstrap_arr)),
        bootstrap_std=float(np.std(bootstrap_arr)),
        ci_95_lower=float(np.percentile(bootstrap_arr, 2.5)),
        ci_95_upper=float(np.percentile(bootstrap_arr, 97.5)),
        n_bootstrap=n_bootstrap,
        seed_hash=seed_hash,
    )

    # Compute drift
    mid_point = n_runs_used // 2
    baseline_values = values[mid_point:]
    recent_values = values[:mid_point]
    drift = compute_drift(baseline_values, recent_values)

    # Get calibration summary
    calibration = await get_calibration_summary(db, tenant_id, node_id)

    # Compute percentiles
    percentiles = {
        "p5": float(np.percentile(arr, 5)),
        "p25": float(np.percentile(arr, 25)),
        "p50": float(np.percentile(arr, 50)),
        "p75": float(np.percentile(arr, 75)),
        "p95": float(np.percentile(arr, 95)),
        "mean": float(np.mean(arr)),
        "std": float(np.std(arr)),
        "min": float(np.min(arr)),
        "max": float(np.max(arr)),
    }

    return ReliabilityDetailResponse(
        status="ok",
        n_runs_total=n_runs_total,
        n_runs_used=n_runs_used,
        sensitivity=sensitivity,
        stability=stability,
        drift=drift,
        calibration=calibration,
        raw_values=values,
        percentiles=percentiles,
        bootstrap_samples=bootstrap_samples,
        audit=audit,
    )
